File: src/PythonClient/src/quixstreams/kafkastreamingclient.py
from typing import Dict, Union

from .configuration import SecurityOptions
from .helpers.enumconverter import EnumConverter as ec
from .helpers.nativedecorator import nativedecorator
from .models import CommitOptions, CommitMode, AutoOffsetReset, CodecType
from .models.netdict import NetDict
from .native.Python.QuixStreamsTelemetry.Kafka.AutoOffsetReset import AutoOffsetReset as AutoOffsetResetInterop
from .native.Python.QuixStreamsStreaming.KafkaStreamingClient import KafkaStreamingClient as sci
from .native.Python.QuixStreamsStreaming.KafkaStreamingClientExtensions import KafkaStreamingClientExtensions as kscei
from .native.Python.QuixStreamsStreaming.Models.CommitMode import CommitMode as CommitModeInterop
from .raw import RawTopicConsumer, RawTopicProducer
from .topicconsumer import TopicConsumer
from .topicproducer import TopicProducer


@nativedecorator
class KafkaStreamingClient(object):
    """
    A Kafka streaming client capable of creating topic consumer and producers.
    """

    def __init__(self, broker_address: str, security_options: SecurityOptions = None, properties: Dict[str, str] = None, debug: bool = False):
        """
        Initializes a new instance of the KafkaStreamingClient.

        Args:
            broker_address: The address of the Kafka cluster.
            security_options: Optional security options for the Kafka client.
            properties: Optional extra properties for broker configuration.
            debug: Whether debugging should be enabled. Defaults to False.
        """

        secu_opts_hptr = None
        if security_options is not None:
            secu_opts_hptr = security_options.get_net_pointer()

        net_properties_hptr = None
        if properties is not None:
            net_properties = NetDict.constructor_for_string_string()
            for key in properties:
                net_properties[key] = properties[key]
            net_properties_hptr = net_properties.get_net_pointer()

        self._interop = sci(sci.Constructor(broker_address, secu_opts_hptr, properties=net_properties_hptr, debug=debug))

    def get_topic_consumer(self, topic: str, consumer_group: str = None, commit_settings: Union[CommitOptions, CommitMode] = None,
                              auto_offset_reset: AutoOffsetReset = AutoOffsetReset.Latest) -> TopicConsumer:
        """
        Gets a topic consumer capable of subscribing to receive incoming streams.

        Args:
            topic: The name of the topic.
            consumer_group: The consumer group ID to use for consuming messages. Defaults to None.
            commit_settings: The settings to use for committing. If not provided, defaults to committing every 5000 messages or 5 seconds, whichever is sooner.
            auto_offset_reset: The offset to use when there is no saved offset for the consumer group. Defaults to AutoOffsetReset.Latest.

        Returns:
            TopicConsumer: An instance of TopicConsumer for the specified topic.
        """

        net_offset_reset = AutoOffsetResetInterop.Latest
        if auto_offset_reset is not None:
            net_offset_reset = ec.enum_to_another(auto_offset_reset, AutoOffsetResetInterop)

        if isinstance(commit_settings, CommitMode):
            net_commit_settings = ec.enum_to_another(commit_settings, CommitModeInterop)

            hptr = kscei.GetTopicConsumer(self._interop.get_interop_ptr__(), topic, consumer_group, net_commit_settings, net_offset_reset)
        else:
            if isinstance(commit_settings, CommitOptions):
                hptr = self._interop.GetTopicConsumer(topic, consumer_group, commit_settings.get_net_pointer(), net_offset_reset)
            else:
                hptr = self._interop.GetTopicConsumer(topic, consumer_group, None, net_offset_reset)

        return TopicConsumer(hptr)

    def get_topic_producer(self, topic: str) -> TopicProducer:
        """
        Gets a topic producer capable of publishing stream messages.

        Args:
            topic: The name of the topic.

        Returns:
            TopicProducer: An instance of TopicProducer for the specified topic.
        """
        hptr = self._interop.GetTopicProducer(topic)
        return TopicProducer(hptr)

    def get_raw_topic_consumer(self, topic: str, consumer_group: str = None, auto_offset_reset: Union[AutoOffsetReset, None] = None) -> RawTopicConsumer:
        """
        Gets a topic consumer capable of subscribing to receive non-quixstreams incoming messages.

        Args:
            topic: The name of the topic.
            consumer_group: The consumer group ID to use for consuming messages. Defaults to None.
            auto_offset_reset: The offset to use when there is no saved offset for the consumer group. Defaults to None.

        Returns:
            RawTopicConsumer: An instance of RawTopicConsumer for the specified topic.
        """

        py_offset_reset = AutoOffsetReset.Earliest
        if auto_offset_reset is not None:
            py_offset_reset = ec.enum_to_another(auto_offset_reset, AutoOffsetResetInterop)

        raw_topic_hptr = self._interop.GetRawTopicConsumer(topic, consumer_group, py_offset_reset)
        return RawTopicConsumer(raw_topic_hptr)

    def get_raw_topic_producer(self, topic: str) -> RawTopicProducer:
        """
        Gets a topic producer capable of publishing non-quixstreams messages.

        Args:
            topic: The name of the topic.

        Returns:
            RawTopicProducer: An instance of RawTopicProducer for the specified topic.
        """
        raw_topic_hptr = self._interop.GetRawTopicProducer(topic)
        return RawTopicProducer(raw_topic_hptr)

File: src/CsharpClient/QuixStreams.Streaming/KafkaStreamingClient.cs
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;
using QuixStreams;
using QuixStreams.Kafka.Transport;
using QuixStreams.Streaming.Configuration;
using QuixStreams.Streaming.Models;
using QuixStreams.Streaming.Raw;
using QuixStreams.Streaming.Utils;
using QuixStreams.Telemetry.Configuration;
using QuixStreams.Telemetry.Kafka;

namespace QuixStreams.Streaming
{
    public interface IKafkaStreamingClient
    {
        /// <summary>
        /// Gets a topic consumer capable of subscribing to receive incoming streams.
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <param name="consumerGroup">The consumer group id to use for consuming messages. If null, consumer group is not used and only consuming new messages.</param>
        /// <param name="options">The settings to use for committing</param>
        /// <param name="autoOffset">The offset to use when there is no saved offset for the consumer group.</param>
        /// <returns>Instance of <see cref="ITopicConsumer"/></returns>
        ITopicConsumer GetTopicConsumer(string topic, string consumerGroup = null, CommitOptions options = null, AutoOffsetReset autoOffset = AutoOffsetReset.Latest);

        /// <summary>
        /// Gets a topic consumer capable of subscribing to receive non-quixstreams incoming messages. 
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <param name="consumerGroup">The consumer group id to use for consuming messages. If null, consumer group is not used and only consuming new messages.</param>
        /// <param name="autoOffset">The offset to use when there is no saved offset for the consumer group.</param>
        /// <returns>Instance of <see cref="IRawTopicConsumer"/></returns>
        IRawTopicConsumer GetRawTopicConsumer(string topic, string consumerGroup = null, AutoOffsetReset? autoOffset = null);

        /// <summary>
        /// Gets a topic producer capable of publishing non-quixstreams messages.  
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <returns>Instance of <see cref="IRawTopicProducer"/></returns>
        IRawTopicProducer GetRawTopicProducer(string topic);

        /// <summary>
        /// Gets a topic producer capable of publishing stream messages. 
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <returns>Instance of <see cref="ITopicProducer"/></returns>
        ITopicProducer GetTopicProducer(string topic);
    }

    /// <summary>
    /// A Kafka streaming client capable of creating topic consumer and producers.
    /// </summary>
    public class KafkaStreamingClient : IKafkaStreamingClient
    {
        private readonly ILogger logger = QuixStreams.Logging.CreateLogger<KafkaStreamingClient>();
        private readonly string brokerAddress;
        private readonly Dictionary<string, string> brokerProperties;
        
        static KafkaStreamingClient()
        {
            // this will guarantee that the codec is set to default, without modifying
            var codec = CodecSettings.CurrentCodec;
        }
        
        /// <summary>
        /// Initializes a new instance of <see cref="KafkaStreamingClient"/>
        /// </summary>
        /// <param name="brokerAddress">Address of Kafka cluster.</param>
        /// <param name="securityOptions">Optional security options.</param>
        /// <param name="properties">Additional broker properties</param>
        /// <param name="debug">Whether debugging should be enabled</param>
        public KafkaStreamingClient(string brokerAddress, SecurityOptions securityOptions = null, IDictionary<string, string> properties = null, bool debug = false)
        {
            this.brokerAddress = brokerAddress;
            if (securityOptions == null)
            {
                this.brokerProperties = new Dictionary<string, string>();
            }
            else
            {
                var securityOptionsBuilder = new SecurityOptionsBuilder();

                if (securityOptions.UseSsl)
                {
                    securityOptionsBuilder.SetSslEncryption(securityOptions.SslCertificates);
                }
                else
                {
                    securityOptionsBuilder.SetNoEncryption();
                }

                if (securityOptions.UseSasl)
                {
                    if (!Enum.TryParse(securityOptions.SaslMechanism.ToString(), true, out Confluent.Kafka.SaslMechanism parsed))
                    {
                        throw new ArgumentOutOfRangeException(nameof(securityOptions.SaslMechanism), "Unsupported sasl mechanism " + securityOptions.SaslMechanism);
                    }

                    securityOptionsBuilder.SetSaslAuthentication(securityOptions.Username, securityOptions.Password, parsed);
                }
                else
                {
                    securityOptionsBuilder.SetNoAuthentication();
                }

                this.brokerProperties = securityOptionsBuilder.Build();
            }

            if (properties != null)
            {
                foreach (var property in properties)
                {
                    this.brokerProperties[property.Key] = property.Value;
                }
            }

            if (debug) this.brokerProperties["debug"] = "all";
        }
        
        /// <summary>
        /// Gets a topic consumer capable of subscribing to receive incoming streams.
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <param name="consumerGroup">The consumer group id to use for consuming messages. If null, consumer group is not used and only consuming new messages.</param>
        /// <param name="options">The settings to use for committing</param>
        /// <param name="autoOffset">The offset to use when there is no saved offset for the consumer group.</param>
        /// <returns>Instance of <see cref="ITopicConsumer"/></returns>
        public ITopicConsumer GetTopicConsumer(string topic, string consumerGroup = null, CommitOptions options = null, AutoOffsetReset autoOffset = AutoOffsetReset.Latest)
        {
            var kafkaReaderConfiguration = new TelemetryKafkaConsumerConfiguration(brokerAddress, consumerGroup, brokerProperties)
            {
                CommitOptions = options,
                AutoOffsetReset = autoOffset.ConvertToKafka()
            };

            var kafkaReader = new TelemetryKafkaConsumer(kafkaReaderConfiguration, topic);

            var topicConsumer = new TopicConsumer(kafkaReader);

            App.Register(topicConsumer);

            return topicConsumer;
        }

        /// <summary>
        /// Gets a topic consumer capable of subscribing to receive non-quixstreams incoming messages. 
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <param name="consumerGroup">The consumer group id to use for consuming messages. If null, consumer group is not used and only consuming new messages.</param>
        /// <param name="autoOffset">The offset to use when there is no saved offset for the consumer group.</param>
        /// <returns>Instance of <see cref="IRawTopicConsumer"/></returns>
        public IRawTopicConsumer GetRawTopicConsumer(string topic, string consumerGroup = null, AutoOffsetReset? autoOffset = null)
        {
            var rawTopicConsumer = new RawTopicConsumer(brokerAddress, topic, consumerGroup, brokerProperties, autoOffset ?? AutoOffsetReset.Latest);

            App.Register(rawTopicConsumer);
            
            return rawTopicConsumer;
        }

        /// <summary>
        /// Gets a topic producer capable of publishing non-quixstreams messages.  
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <returns>Instance of <see cref="IRawTopicProducer"/></returns>
        public IRawTopicProducer GetRawTopicProducer(string topic)
        {
            var rawTopicProducer = new RawTopicProducer(brokerAddress, topic, brokerProperties);

            App.Register(rawTopicProducer);
            return rawTopicProducer;
        }
        
        /// <summary>
        /// Gets a topic producer capable of publishing stream messages. 
        /// </summary>
        /// <param name="topic">Name of the topic.</param>
        /// <returns>Instance of <see cref="ITopicProducer"/></returns>
        public ITopicProducer GetTopicProducer(string topic)
        {
            var topicProducer = new TopicProducer(new KafkaProducerConfiguration(brokerAddress, brokerProperties), topic);
            
            App.Register(topicProducer);

            return topicProducer;
        }
    }

    /// <summary>
    /// Extensions for Streaming Client class
    /// </summary>
    public static class KafkaStreamingClientExtensions
    {
        /// <summary>
        /// Open an topic consumer capable of subscribing to receive incoming streams
        /// </summary>
        /// <param name="client">Streaming Client instance</param>
        /// <param name="topic">Name of the topic.</param>
        /// <param name="autoOffset">The offset to use when there is no saved offset for the consumer group.</param>
        /// <param name="consumerGroup">The consumer group id to use for consuming messages. If null, consumer group is not used and only consuming new messages.</param>
        /// <param name="commitMode">The commit strategy to use for this topic</param>
        /// <returns>Instance of <see cref="ITopicConsumer"/></returns>
        public static ITopicConsumer GetTopicConsumer(this IKafkaStreamingClient client, string topic, string consumerGroup = null, CommitMode commitMode = CommitMode.Automatic, AutoOffsetReset autoOffset =  AutoOffsetReset.Latest)
        {
            switch (commitMode)
            {
                case CommitMode.Automatic:
                    return client.GetTopicConsumer(topic, consumerGroup, autoOffset: autoOffset);
                case CommitMode.Manual:
                    var commitOptions = new CommitOptions()
                    {
                        AutoCommitEnabled = false
                    };
                    return client.GetTopicConsumer(topic, consumerGroup, commitOptions, autoOffset);
                default:
                    throw new ArgumentOutOfRangeException(nameof(commitMode), commitMode, null);
            }
        }
    }
}
